### PR TITLE
[autoscaler] Support TPU as Head Node

### LINKED ISSUE #32052: [Autoscaler][GCP] TPU should be allowed as Head Node
### What happened + What you expected to happen

Autoscaler doesn't allow TPU nodes to act as head nodes

This code is in the `gcp/config.py`
```python
if _is_head_node_a_tpu(config):
  raise RuntimeError("TPUs are not supported as head nodes.")
```

https://github.com/ray-project/ray/blob/45ffe6eb99d96488fdec187bb47a4a78d9b5ee92/python/ray/autoscaler/_private/gcp/config.py#L290

Last I found was here: https://github.com/ray-project/ray/pull/17278/files#r677939519

When the original PR came in from @ijrsvt in July 2021 this didnt work, but now ~18 months later, as far as my tests show this should work just fine now. 

### Versions / Dependencies

Ray version: This is on master (v2.2 and v3) show this
TPU VM v2-8 on GCP

### Reproduction script

By manually making a service account on a TPU VM node with the same permissions, I was able to directly call and create a TPU worker node

```
$ cloud auth activate-service-account  ...

$ gcloud alpha compute tpus tpu-vm create ray-worker-12312df --version='v2-alpha' --zone=us-central1-f --accelerator-type="v2-8" --preemptible
Create request issued for: [ray-worker-12312df]
Waiting for operation [projects/approx-labs-research/locations/us-ce
ntral1-f/operations/operation-1675097649815-5f37e10619ee9-7df9192e-b
4985a36] to complete...done.                                        
Created tpu [ray-worker-12312df].
```

### Issue Severity

None

### PR DESCRIPTION
**Pull request by bluecoconut**
## Why are these changes needed?

Code was blocking using TPU as head node. Not sure about historical scopes issues that may have been resolved on GCP Side.

Separate from those, there was an issue where the `serviceAccounts` key (and `scopes` sub-key) in the `node` part of the create VM API are different on the TPU endpoint (seems like this is a GCP consistency issue... and i don't feel like chasing this upstream right now). This difference means that you have to change `serviceAccounts` -> `serviceAccount` and only supply 1 for TPU head node. 

With these changes, it is now possible to have a homogenous cluster of all TPU nodes.

## Related issue number

Closes https://github.com/ray-project/ray/issues/32052

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [x] This PR is not tested :(

Screenshots showing it working
image
image
image

### CODE REVIEW

**Comment by wuisawesome** [comment]
@bluecoconut thanks for the contribution!

@Yard1 Maybe you can do an initial pass or provide some high level thoughts on this? I believe you did the previous work on this (#17278)

**Comment by Yard1** [comment]
High level this looks fine to me, the only reason it was disabled before was because it was impossible to start nodes from another TPU node. It looks like that has been resolved on Google's side, so there's no reason not to go ahead with this PR!

File: python/ray/autoscaler/gcp/tpu.yaml
Comment (wuisawesome) [inline python/ray/autoscaler/gcp/tpu.yaml:26]:
```suggestion
        min_workers: 1
```

nit

**Review by wuisawesome — CHANGES_REQUESTED** [review]
I left a few minor comments. We don't have a good way of unit testing this in CI yet, but do you mind posting the "repro" for how you are currently testing this locally? Maybe either in this issue or adding it as a comment at the top of `tpu.yaml`? 

I see a ` python3 -c "import jax; jax.device_count(); jax.numpy.add(1, 1)"` in setup commands. 

Should we also try running that from a remote task or something after the cluster has started? 

If I'm not mistaken presumably something like 

```
import ray

@ray.remote(resources={"TPU": 1})
class Foo:
  def foo():
    assert jax.device_count() == 1
    assert jax.numpy.add(1, 1) == 2


actors = [Foo.remote() for _ in range(2)]
ray.get([actor.foo.remote() for actor in actors])
```

**Comment by wuisawesome** [comment]
cc @architkulkarni

File: python/ray/autoscaler/gcp/tpu.yaml
Comment (bluecoconut) [reply python/ray/autoscaler/gcp/tpu.yaml:52]: Sure

**Comment by bluecoconut** [comment]
> We don't have a good way of unit testing this in CI yet, but do you mind posting the "repro" for how you are currently testing this locally? 

Yeah, I'm simply running `ray up cluster.yaml`, and then `ray attach` and using it... haha. 

It went from erroring immediately (on `ray up`) to functioning as a cluster would. Not sure there's too much else there to describe.

In theory cpuld include a script, but I assume any actual user of this will heavily modify the startup commands and such (I know I am, my local cluster.yaml is quite different from the saved one here). 

This was roughly an attempt to keep the same previous `tpu.yaml` which included a reference to jax-swarm (which, I never actually got working, due to version issues with packages)

File: python/ray/autoscaler/_private/gcp/node.py
Comment (bluecoconut) [reply python/ray/autoscaler/_private/gcp/node.py:690]: I actually switched machines recently, and haven't set up ray in development mode. 

Trying to quickly hack on it (right now) led to some opaque errors when trying to run tests (missing imports) likely due to how i was actually editing the code in site-packages rather than in a proper `pip install -e .` style and master being ahead by enough... 

Also, i'm on a M1 mac, and it seems like that's not supported in the nightlies, which is described as the way to set up in python only development mode: 
https://docs.ray.io/en/latest/ray-contribute/development.html#building-ray-python-only
 
> At this time, Apple Silicon ray wheels are being published for releases only. As support stabilizes, nightly wheels will be published in the future.
https://docs.ray.io/en/latest/ray-overview/installation.html#m1-mac-apple-silicon-support

-- All this said, I think I am not going to fight all of the above and figure out how to install ray from source on M1 right now just to add this test. I'm obviously open to anyone else contributing and adding the test who has the environment set up.

I'll loop back around to this in a few weeks or so (depending on if I have free time) and will hack on getting a dev environment set up or set up in some quick linux VM to add it.

**Comment by richardliaw** [comment]
@architkulkarni can you review this when you get the chance?

cc @Yard1

**Review by architkulkarni — APPROVED** [review]
I think this is good to go, ideally we should have a unit test for the new logic.

@bluecoconut Since you've submitted the PR we've added https://github.com/ray-project/ray/blob/master/python/ray/autoscaler/launch_and_verify_cluster.py to our release tests.  Do you mind running it in your account to see if it passes?  Once your PR is merged we can add a release test on our side to automatically run it. (We'll need to set up permissions on our side so I don't think it's feasible for you to contribute that part)

**Comment by bluecoconut** [comment]
@architkulkarni 

Just did this and it worked, node turned on, installed things, and then turned off. To help remove confusion and offer a better starting point, I simplified the `tpu.yaml` as well. 

Note: It actually spun up and down correctly, but the script itself failed on the spin down because it takes a very long time for TPU nodes to spin up and down (~2-4 minutes): I got a timeout error there.

**Review by wuisawesome — APPROVED** [review]
I suppose I don't want to block merging this since it's an overall positive contribution, but I'm definitely a little scared of the lack of testing in the GCP node provider right now (I almost think the only reason this thing isn't breaking all the time is that we don't touch it).

A compromise I'd propose: @architkulkarni do you mind manually verifying this change works for now, and file some follow up tracking issues for unit testing this portion of the code and e2e testing this TPU yaml, and we can triage the priorities of those?

File: python/ray/autoscaler/gcp/tpu.yaml
Comment (architkulkarni) [inline python/ray/autoscaler/gcp/tpu.yaml:43]: I think this comment is useful, shouldn't we keep it?

**Review by architkulkarni — COMMENTED** [review]
I think it might be better to maintain two yamls, a `tpu-minimal.yaml` alongside the existing `tpu-swarm-jax.yaml`, because I think the existing yaml might still be useful to people.  @Yard1 @bluecoconut what do you think?

**Comment by architkulkarni** [comment]
Kicked off existing GCP release tests here https://buildkite.com/ray-project/release-tests-pr/builds/38632#_

**Comment by architkulkarni** [comment]
Yeah got a timeout as well, but on `ray up`:

```
======================================
Starting new cluster...
Cluster: tputest

2023-05-17 12:14:03,220 INFO util.py:376 -- setting max workers for head node type to 0
2023-05-17 12:14:03,220 INFO util.py:380 -- setting max workers for ray_tpu to 7
Loaded cached provider configuration from /var/folders/75/7kpb8gj57k11df4ksj9lnt900000gn/T/ray-config-d0ba6b245a69e13250a63c3df097d59f80af8822
If you experience issues with the cloud provider, try re-running the command with --no-config-cache.
No head node found. Launching a new cluster. Confirm [y/N]: y [automatic, due to --yes]

Usage stats collection is enabled. To disable this, add `--disable-usage-stats` to the command that starts the cluster, or run the following command: `ray disable-usage-stats` before starting the cluster. See https://docs.ray.io/en/master/cluster/usage-stats.html for more details.

Acquiring an up-to-date head node
2023-05-17 12:14:10,182 INFO node.py:563 -- wait_for_tpu_operation: Waiting for operation projects/archit-tpu-partner/locations/us-central2-b/operations/operation-1684350848919-5fbe87e06e626-f34bc5ac-557fbd8d to finish...
Traceback (most recent call last):
  File "/Users/archit/anaconda3/envs/ray-py38/bin/ray", line 33, in <module>
    sys.exit(load_entry_point('ray', 'console_scripts', 'ray')())
  File "/Users/archit/ray/python/ray/scripts/scripts.py", line 2456, in main
    return cli()
  File "/Users/archit/anaconda3/envs/ray-py38/lib/python3.8/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/Users/archit/anaconda3/envs/ray-py38/lib/python3.8/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/Users/archit/anaconda3/envs/ray-py38/lib/python3.8/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/archit/anaconda3/envs/ray-py38/lib/python3.8/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/archit/anaconda3/envs/ray-py38/lib/python3.8/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/Users/archit/ray/python/ray/autoscaler/_private/cli_logger.py", line 856, in wrapper
    return f(*args, **kwargs)
  File "/Users/archit/ray/python/ray/scripts/scripts.py", line 1270, in up
    create_or_update_cluster(
  File "/Users/archit/ray/python/ray/autoscaler/_private/commands.py", line 282, in create_or_update_cluster
    get_or_create_head_node(
  File "/Users/archit/ray/python/ray/autoscaler/_private/commands.py", line 726, in get_or_create_head_node
    provider.create_node(head_node_config, head_node_tags, 1)
  File "/Users/archit/ray/python/ray/autoscaler/_private/gcp/node_provider.py", line 46, in method_with_retries
    return method(self, *args, **kwargs)
  File "/Users/archit/ray/python/ray/autoscaler/_private/gcp/node_provider.py", line 178, in create_node
    results = resource.create_instances(
  File "/Users/archit/ray/python/ray/autoscaler/_private/gcp/node.py", line 286, in create_instances
    results = [
  File "/Users/archit/ray/python/ray/autoscaler/_private/gcp/node.py", line 287, in <listcomp>
    (self.wait_for_operation(operation), node_name)
  File "/Users/archit/ray/python/ray/autoscaler/_private/gcp/node.py", line 570, in wait_for_operation
    self.resource.projects()
  File "/Users/archit/anaconda3/envs/ray-py38/lib/python3.8/site-packages/googleapiclient/_helpers.py", line 130, in positional_wrapper
    return wrapped(*args, **kwargs)
  File "/Users/archit/anaconda3/envs/ray-py38/lib/python3.8/site-packages/googleapiclient/http.py", line 923, in execute
    resp, content = _retry_request(
  File "/Users/archit/anaconda3/envs/ray-py38/lib/python3.8/site-packages/googleapiclient/http.py", line 222, in _retry_request
    raise exception
  File "/Users/archit/anaconda3/envs/ray-py38/lib/python3.8/site-packages/googleapiclient/http.py", line 191, in _retry_request
    resp, content = http.request(uri, method, *args, **kwargs)
  File "/Users/archit/anaconda3/envs/ray-py38/lib/python3.8/site-packages/google_auth_httplib2.py", line 218, in request
    response, content = self.http.request(
  File "/Users/archit/anaconda3/envs/ray-py38/lib/python3.8/site-packages/httplib2/__init__.py", line 1724, in request
    (response, content) = self._request(
  File "/Users/archit/anaconda3/envs/ray-py38/lib/python3.8/site-packages/httplib2/__init__.py", line 1444, in _request
    (response, content) = self._conn_request(conn, request_uri, method, body, headers)
  File "/Users/archit/anaconda3/envs/ray-py38/lib/python3.8/site-packages/httplib2/__init__.py", line 1396, in _conn_request
    response = conn.getresponse()
  File "/Users/archit/anaconda3/envs/ray-py38/lib/python3.8/http/client.py", line 1348, in getresponse
    response.begin()
  File "/Users/archit/anaconda3/envs/ray-py38/lib/python3.8/http/client.py", line 316, in begin
    version, status, reason = self._read_status()
  File "/Users/archit/anaconda3/envs/ray-py38/lib/python3.8/http/client.py", line 277, in _read_status
    line = str(self.fp.readline(_MAXLINE + 1), "iso-8859-1")
  File "/Users/archit/anaconda3/envs/ray-py38/lib/python3.8/socket.py", line 669, in readinto
    return self._sock.recv_into(b)
  File "/Users/archit/anaconda3/envs/ray-py38/lib/python3.8/ssl.py", line 1241, in recv_into
    return self.read(nbytes, buffer)
  File "/Users/archit/anaconda3/envs/ray-py38/lib/python3.8/ssl.py", line 1099, in read
    return self._sslobj.read(len, buffer)
socket.timeout: The read operation timed out
Traceback (most recent call last):
  File "launch_and_verify_cluster.py", line 218, in <module>
    run_ray_commands(cluster_config, retries, no_config_cache)
  File "launch_and_verify_cluster.py", line 124, in run_ray_commands
    subprocess.run(cmd, check=True)
  File "/Users/archit/anaconda3/envs/ray-py38/lib/python3.8/subprocess.py", line 516, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['ray', 'up', '-v', '-y', '/var/folders/75/7kpb8gj57k11df4ksj9lnt900000gn/T/tmpbpmy66es.yaml']' returned non-zero exit status 1.
```

Later I went to the cloud console and saw that the `tpu-head` was successfully created. @bluecoconut is this the same kind of timeout you saw as well?

**Comment by bluecoconut** [comment]
@architkulkarni no my error was on the cleanup part. It successfully made the head node, did some rsync and ran the commands, and then the resolved state was actually a cleaned up machine, but it just took ~5-10 minutes for the machine to actually successfully delete, and the script timed out on the local client on its cleanup step.

**Comment by architkulkarni** [comment]
I think my timeout error was transient.  I got a different error this time:

```
  [6/7] Running setup commands
    (0/4) sudo apt install python-is-python3 -y
    Running `sudo apt install python-is-python3 -y`
To run a command as administrator (user "root"), use "sudo <command>".
See "man sudo_root" for details.

Reading package lists... Done
Building dependency tree       
Reading state information... Done
Package python-is-python3 is not available, but is referred to by another package.
This may mean that the package is missing, has been obsoleted, or
is only available from another source
However the following packages replace it:
  python-is-python2

E: Package 'python-is-python3' has no installation candidate
```

@bluecoconut can you take a look? Is it necessary here? (I don't see it in any other setup YAML files)

**Comment by bluecoconut** [comment]
It was necessary for me. But, also worth noting my actual clusters have very custom setup commands, that is different from this. This setup script/yaml was just the minimum I got to work on the `v2-alpha` vm base machines specifically. Did you change the machine type or image in your run?

**Comment by bluecoconut** [comment]
Also I'll note, I vaguely remember (3 months ago), I think this was necessary because ray attach or ray submit or something assumed it could run "python" as the command to execute a python script on the node (seemed to be hard coded there to find python interpreter that way). So, I included this since the machine was using python3 as the command, and I didn't set up any virtualenv stuff. So this was fastest way to make the other things work.

**Comment by architkulkarni** [comment]
I see, thanks for the additional info.  I was using 

```
            acceleratorType: v4-8
            runtimeVersion: tpu-vm-v4-base
```
since I think that's all I have enabled in my account currently.  I'll double check and see if I can try again with `v2-alpha`.

**Comment by guseggert** [comment]
Hi, wondering if there are any updates here? Anything I can help with to get this over the line?

File: python/ray/autoscaler/gcp/tpu.yaml
Comment (architkulkarni) [inline python/ray/autoscaler/gcp/tpu.yaml:30]:
```suggestion
    availability_zone: us-central1-b
```

I couldn't get it to work with `us-central1-f`.  The docs say "Note: Zone us-central1-f is available only to [TPU Research Cloud (TRC) participants](https://sites.research.google/trc)."  So perhaps `b` will be better for more users.

**Review by architkulkarni — COMMENTED** [review]
I was able to successfully run `ray up tpu.yaml` and after waiting a while, `ray status` on the head node returned successfully showing the TPU head and the TPU worker:
Screenshot 2023-06-08 at 3 11 25 PM

My only concern before merging is that this PR seems to be deleting the existing Jax example.  I think we should keep both, and add the new YAML as a new file `tpu-minimal.yaml` or `tpu-head.yaml`.  @Yard1 would you agree, or is the existing Jax example not useful anymore?

File: python/ray/autoscaler/gcp/tpu.yaml
Comment (architkulkarni) [reply python/ray/autoscaler/gcp/tpu.yaml:30]: (doc link) https://cloud.google.com/tpu/docs/regions-zones#overview

**Comment by Yard1** [comment]
Yeah we can get rid of the Jax example.

**Comment by architkulkarni** [comment]
> Yeah we can get rid of the Jax example.

Sounds good, merging as is.

**Comment by architkulkarni** [comment]
The [existing GCP release tests passed](https://buildkite.com/ray-project/release-tests-pr/builds/38632#) .

**Pull request merged**
